File: Python/shortest_path.py
import math
from queue import PriorityQueue

while True:
    n, m, q, s = map(int, input().split())
    if n == m == q == s == 0:
        exit()
    dp = [math.inf for _ in range(n)]
    dp[s] = 0
    edges = [[0 for _ in range(n)] for _ in range(n)]
    key = [[] for _ in range(n)]
    visited = [False for _ in range(n)]

    for _ in range(m):
        a, b, w = map(int, input().split())
        edges[a][b] = w
        key[a].append(b)

    pq = PriorityQueue()
    curr = s
    while unvisited:
        if curr in unvisited: unvisited.remove(curr)
        for neighbour in edges[curr]:
            dp[neighbour] = min(dp[neighbour], dp[curr] + edges[curr][neighbour])

        best = math.inf
        nxt = None
        for node in unvisited:
            if dp[node] < best:
                best = dp[node]
                nxt = node
        if nxt == None: break
        curr = nxt
        
    
    for _ in range(q):
        e = int(input())
        if dp[e] == math.inf:
            print("Impossible")
        else:
            print(dp[e])
            

File: C++/shortest_path.cpp
#include <iostream>
#include <math.h>
#include <map>
#include <set>
#include <queue>

using namespace std;

const int INF = 1000000000;

int main() {
    while (true) {

        int n, m, q, s;
        cin >> n >> m >> q >> s;

        if (n == 0 && m == 0 && q == 0 && s == 0) {
            return 0;
        }

        int dp[n];
        bool visited[n];
        for (int i = 0; i < n; i++) dp[i] = INF;
        for (int i = 0; i < n; i++) visited[i] = false;
        dp[s] = 0;

        int edges[n][n];
        auto edgs = vector<vector<int> >(n, vector<int>(n, 0));
        vector<int> key[n];

        for (int i = 0; i < m; i++) {
            int a, b, w;
            cin >> a >> b >> w;
            if (edgs[a][b] == 0) {
                edges[a][b] = w;
                edgs[a][b] = 1;
                key[a].push_back(b);
            }
            else {
                edges[a][b] = min(edges[a][b], w);
            }

        }

        priority_queue<pair<int, int>, vector<pair<int, int> >, greater<pair<int, int> > > pq;


        int curr = s;
        bool run = true;
        while (run) {
            visited[curr] = true;

            for (int neighbour : key[curr]) {
                dp[neighbour] = min(dp[neighbour], dp[curr] + edges[curr][neighbour]);
                pq.push(make_pair(dp[neighbour], neighbour));
            }

            try {
                while (visited[curr]) {
                    if (pq.size() == 0) {
                        run = false;
                        break;
                    }
                    curr = pq.top().second;
                    pq.pop();
                }   
            }
            catch (exception e) {
                break;
            }
            
        }

        for (int i = 0; i < q; i++) {
            int e;
            cin >> e;
            int val = dp[e];
            if (val == INF) {
                cout << "Impossible" << endl;
            } else {
                cout << val << endl;
            }
        }
    }

    return 0;
}